kernelci/build.py

```diff
@@ -492,6 +492,512 @@ def _run_make(kdir, arch, target=None, jopt=None, silent=True, cc='gcc',
     return shell_cmd(cmd, True)
 
 
+class Step:
+    """Kernel build step"""
+
+    def __init__(self, kdir, output_path=None, log=None, reset=False):
+        """Each Step deals with a part of the build and its related meta-data
+
+        This abstract class handles the common code to run any kernel build
+        step, such as running `make`, managing log files or checking whether
+        some kernel config options are enabled.  It uses the `bmeta.json` file
+        to share meta-data between each step and later on when sending the
+        build to a database.
+
+        Each concrete class should implement the `run()` method, with arbitrary
+        arguments to perform their dedicated tasks.
+
+        *kdir* is the path to the kernel source tree directory
+        *output_path* is the path to the build output directory
+        *log* is the name of the log file within the output directory
+        *reset* is whether the meta-data should be reset in this step
+        """
+        self._kdir = kdir
+        self._output_path = output_path or os.path.join(kdir, 'build')
+        self._bmeta_path = os.path.join(self._output_path, 'bmeta.json')
+        self._steps_path = os.path.join(self._output_path, 'steps.json')
+        self._log_path = log
+        self._bmeta = dict()
+        self._steps = list()
+        self._dot_config = None
+        self._start_time = time.time()
+        if not os.path.exists(self._output_path):
+            os.mkdir(self._output_path)
```

Comment:
Is it intentional to create only the top level directory if it doesn't exist and fail if other parts of the path are missing?

Reply:
Yes, only the output directory has to be there because that's where the files created by this module will be stored, with no sub-directories.  Anything added in sub-directories within there will be created by the kernel build system.